File: R/st-density.R
# Isoband to raster: https://gist.github.com/mdsumner/a06faedbdf30b6d5c1820487f81a6959

#' Compute 2D density
#'
#' Choose from methods \code{"kde2d"} and \code{"bkde2D"},
#' and various return geometries
#'
#' @param x (sf/sfc) [missing] Spatial data
#' @param return_geometry [\code{"point"}] (character) What gets returned?
#' @param method (character) [\code{"kde2d"}] How should density be computed?
#' @param bw (numeric) [\code{NULL}] Binwidth
#' @param n (numeric) [\code{NULL}] Grid size
#' @param bins (numeric) [\code{NULL}] Number of contour bins
#' @param truncate (logical) [\code{TRUE}] See \code{KernSmooth::bkde2D}
#' @param x_expansion,y_expansion (numeric) [\code{NULL}] Expansion multiple
#'   applied to x-/y-range and used in \code{lims} and \code{range.x}
#'   arguments for methods \code{"kde2d"} and \code{"bkde2D"}, respectively.
#' @param levels_high,levels_low (numeric) Low/high breakpoints
#'  for isolines and isobands.
#'
#' @name st_density
#' @export
st_density <- function (x,
                        return_geometry = "point",
                        method = "kde2d",
                        bw = NULL,
                        n = 200,
                        bins = NULL,
                        truncate = TRUE,
                        x_expansion = NULL,
                        y_expansion = NULL,
                        levels_low = NULL,
                        levels_high = NULL) {

  UseMethod("st_density")
}

#' @name st_density
#' @export
st_density.sfc <- function (x,
                            return_geometry = "point",
                            method = "kde2d",
                            bw = NULL,
                            n = NULL,
                            bins = NULL,
                            truncate = TRUE,
                            x_expansion = NULL,
                            y_expansion = NULL,
                            levels_low = NULL,
                            levels_high = NULL) {

  if (!inherits(x, "sf")) stop("[st_density]: 'x' must inherit 'sf'.")
  data_coords <- sf::st_coordinates(x)
  x_crs <- sf::st_crs(x)

  if (is.null(n)) {
    n <- switch(method,
                kde2d = 200,
                bkde2D = c(51, 51))
  }

  # Compute density by method and return geometry
  dens <- switch(method,
                 kde2d = sf_compute_kde2d(data_coords, return_geometry, bw, n,
                                          x_expansion, y_expansion),

                 bkde2D = sf_compute_bkde2D(data_coords, return_geometry,
                                            bw, n, x_expansion, y_expansion,
                                            truncate))

  switch(return_geometry,

          point  = {
            x$density  <- dens$density
            x$ndensity <- dens$ndensity
          },

          grid = {
            x <- sf::st_as_sf(dens, coords = c("x", "y"), crs = x_crs)
          },

          raster = {
            # also: https://gis.stackexchange.com/questions/265094/r-grid-of-polygons-to-raster-file-for-esri
            # if (length(n) < 2) n <- rep(n, 2)
            # x <- stars::st_rasterize(dens, nx = n[1], ny = n[2])
            x <- dens %>%
              sf::st_as_sf(coords = c("x", "y"), crs = x_crs) %>%
              stars::st_rasterize() # specify nx/ny?
          },

          polygon = {
            # go to isoband
          },

          isoband = {
            if (is.null(levels_high) | is.null(levels_low)) {
              levels_low  <- .05 * (0:20)
              levels_high <- .05 * (1:21)
            }
            iso_matrix <- tapply(dens$ndensity, dens[, c("y", "x")], identity)
            iso_band <- isoband::isobands(unique(dens$x),
                                          unique(dens$y),
                                          iso_matrix,
                                          levels_low = levels_low,
                                          levels_high = levels_high)
            iso_band <- isoband::iso_to_sfg(iso_band)
            x <- sf::st_sf(level = seq_len(length(iso_band)),
                           crs = x_crs,
                           geometry = sf::st_sfc(iso_band))
          })

  x

}

#' @name st_density
#' @export
st_density.sf <- st_density.sfc

File: Py/st_density.py
"""
2D spatial density estimation for Python

Translates R sfx::st_density functionality using GeoPandas and scikit-learn
"""

import numpy as np
import pandas as pd
import geopandas as gpd
from sklearn.neighbors import KernelDensity
from scipy.stats import gaussian_kde
from scipy.spatial.distance import pdist
import warnings

# Import helper functions
from .density_compute import _compute_kde2d, _compute_bkde2d


def st_density(x, return_geometry="point", method="kde2d", bw=None, n=200, 
               bins=None, truncate=True, x_expansion=None, y_expansion=None,
               levels_low=None, levels_high=None):
    """
    Compute 2D density
    
    Choose from methods 'kde2d' and 'bkde2D', and various return geometries
    
    Parameters
    ----------
    x : GeoDataFrame or GeoSeries
        Spatial data
    return_geometry : str, default "point"  
        What gets returned ("point", "grid", "raster", "polygon", "isoband")
    method : str, default "kde2d"
        How should density be computed ("kde2d", "bkde2D")
    bw : float or array-like, optional
        Binwidth/bandwidth
    n : int or array-like, default 200
        Grid size  
    bins : int, optional
        Number of contour bins
    truncate : bool, default True
        See KernelDensity truncate parameter
    x_expansion : float, optional
        Expansion multiple applied to x-range
    y_expansion : float, optional  
        Expansion multiple applied to y-range
    levels_low : array-like, optional
        Low breakpoints for isolines and isobands
    levels_high : array-like, optional
        High breakpoints for isolines and isobands
        
    Returns
    -------
    GeoDataFrame or DataFrame
        Density estimates with geometry or coordinates
    """
    
    if not isinstance(x, (gpd.GeoDataFrame, gpd.GeoSeries)):
        raise ValueError("x must be a GeoDataFrame or GeoSeries")
        
    # Extract coordinates
    if hasattr(x, 'geometry'):
        coords = np.column_stack([x.geometry.x, x.geometry.y])
    else:
        coords = np.column_stack([x.x, x.y])
    
    x_crs = x.crs if hasattr(x, 'crs') else None
    
    if n is None:
        n = 200 if method == "kde2d" else [51, 51]
    
    # Compute density by method
    if method == "kde2d":
        result = _compute_kde2d(coords, return_geometry, bw, n, 
                               x_expansion, y_expansion)
    elif method == "bkde2D":
        result = _compute_bkde2d(coords, return_geometry, bw, n,
                                x_expansion, y_expansion, truncate)
    else:
        raise ValueError(f"Unknown method: {method}")
    
    # Handle return geometry
    if return_geometry == "point":
        # Add density columns to original data
        if isinstance(x, gpd.GeoDataFrame):
            x = x.copy()
            x['density'] = result['density']  
            x['ndensity'] = result['ndensity']
            return x
        else:
            df = pd.DataFrame({
                'x': coords[:, 0],
                'y': coords[:, 1], 
                'density': result['density'],
                'ndensity': result['ndensity']
            })
            return gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df.x, df.y), crs=x_crs)
            
    elif return_geometry == "grid":
        return gpd.GeoDataFrame(result, 
                               geometry=gpd.points_from_xy(result.x, result.y), 
                               crs=x_crs)
    
    elif return_geometry == "raster":
        # Convert to raster format (could use rasterio/xarray)
        return result  # Simplified for now
        
    elif return_geometry == "isoband":
        if levels_high is None or levels_low is None:
            levels_low = 0.05 * np.arange(21)
            levels_high = 0.05 * np.arange(1, 22)
        
        # Create contour polygons using matplotlib/skimage
        # Simplified implementation
        return result
        
    else:
        raise ValueError(f"Unknown return_geometry: {return_geometry}")